Add convergence criteria to linear solver

nilsvu opened the pull request:
## Proposed changes

Convergence criteria for the linear solver implemented so far:
- Max. number of iterations
- Absolute residual
- Residual relative to the initial residual

Also adds a compute item for `LinearSolver::Tags::HasConverged` that uses the convergence criteria.

Please see the commit messages for more details.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

nilsvu commented on src/NumericalAlgorithms/LinearSolver/Tags.hpp:
```diff
@@ -75,7 +83,7 @@ struct HasConverged : db::SimpleTag {
     return "LinearSolverHasConverged";
   }
-  using type = bool;
+  using type = std::pair<bool, std::string>;
```
Note to self: update docs for this Tag

hrueter commented on src/NumericalAlgorithms/LinearSolver/ConvergenceCriteria.hpp:
```diff
@@ -0,0 +1,110 @@
+  struct RelativeResidual {
+    using type = double;
+    static constexpr OptionString help = {
+        "The residual has decreased by this factor."};
```
The option descriptions in this function are confusing.

E.g. here I think you mean something like
"The solver terminates when the residual decreased by this factor."

The past tense in your text would imply, that this condition is somehow already met at the start of the iteration.

nilsvu replied:
The option strings must be very short, so I chose this style of documentation to complement the option string of the struct. I'm not convinced to change it yet 😄

wthrowe commented:
I haven't looked at the code in detail, but have you considered doing this using the events-and-triggers infrastructure?

nilsvu commented:
@wthrowe I have not, indeed I thought the events-and-triggers are not quite ready for use yet. For this particular purpose (linear solver convergence criteria) they are probably not applicable, except perhaps to replace the `TerminateIfConverged` action. That's because the convergence happens on a singleton parallel component (the residual monitor) in a reduction action, and not on the elements in an algorithm action.

wthrowe commented:
Events and triggers have been ready for use since last February, but they were mostly intended for use in an action iteration sequence so you may be right that they can't be used here.

nilsvu committed:
Add compute item for LinearSolver::Magnitude from its square

This is useful for the conjugate gradient linear solver
that stores the residual magnitude square for its algorihtm,
but needs to take the square-root for residual reporting
and convergence checks.

nilsvu committed:
Have Poisson input files tests only run for 1 iteration

nilsvu commented:
I split the changes into more descriptive commits for easier review

kidder commented:
@nilsleiffischer this fails to compile

nilsvu committed:
Test that linear solver algorithms indeed converge for simple problems

nilsvu committed:
Add convergence criteria for linear solver

nilsvu committed:
Add tags for convergence criteria

Also adds a compute item for HasConverged that uses the convergence criteria

nilsvu committed:
Add convergence criteria to the conjugate gradient linear solver

Also removes the residual magnitude from the elements, since it was
only used for convergence checks and observing.

nilsvu committed:
Add convergence criteria to the GMRES linear solver

Also removes:
- The residual magnitude from the elements, since it was
only used for convergence checks and observing
- The source magnitude from the residual monitor, since it is presently
unused

nilsvu commented:
Made some more changes, in particular added a struct that represents the convergence status instead of a `std::pair<bool, std::string>`.

This is now ready for review. @nilsdeppe we discussed this so perhaps you could review, or forward to someone else? @hrueter would you also have time to review?

nilsdeppe commented:
@nilsleiffischer sure, but it still doesn't compile on Travis. I'm going to wait for it to be compiling and passing tests before review. Marking as "in progress" for now until that's fixed :)

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Convergence.hpp:
```diff
@@ -0,0 +1,169 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines DataBox tags for the linear solver
```
Remove.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Convergence.hpp:
```diff
@@ -0,0 +1,169 @@
+ * - MaxIterations: Matches if the number of iterations exceeds this limit.
+ * - AbsoluteResidual: Matches if the residual has reached this magnitude.
+ * Bounded below by the product of the machine epsilon and the conditioning
+ * number of the linear operator that is being inverted.
```
I don't understand what the second sentence ("Bounded below...") is trying to tell me.

nilsvu replied:
I added more documentation. Does this make it more clear?

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Convergence.hpp:
```diff
@@ -0,0 +1,169 @@
+/*!
+ * \brief Criteria that determine the linear solve has converged
+ *
+ * \details The following criteria are implemented:
```
Do you want to specify which one "wins" if multiple criteria are satisfied on the same iteration?

wthrowe commented on tests/Unit/NumericalAlgorithms/LinearSolver/Test_Convergence.cpp:
```diff
@@ -0,0 +1,123 @@
+
+  {
+    INFO("HasNotConverged");
+    CHECK_FALSE(LinearSolver::HasConverged{});
```
Test serialization of a default-constructed object.

wthrowe commented on tests/InputFiles/Poisson/Input2D.yaml:
```diff
@@ -12,14 +12,19 @@ DomainCreator:
     UpperBound: [3.141592653589793, 3.141592653589793]
     IsPeriodicIn: [false, false]
-    InitialRefinement: [1, 1]
+    InitialRefinement: [0, 0]
```
Was this change intentional?  It seems unrelated to the rest of the commit.

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp:
```diff
@@ -58,51 +54,32 @@ struct InitializeResidualMagnitude {
+    if (UNLIKELY(has_converged and
+                 static_cast<int>(get<::Tags::Verbosity>(box)) >=
+                     static_cast<int>(::Verbosity::Quiet))) {
+      Parallel::printf(MakeString{} << has_converged);
```
`Parallel::printf("%s", has_converged)`  There are several other similar bits of code in this PR, which should also be changed.

The important part of this change is the `"%s"`.  The fact that you can (I think) avoid the `MakeString` is a side benefit.  If the string form of `has_converged` happens to contain a "%" then the current code may segfault.  The first argument to `printf` should (almost) always be a string literal.  (This is GCC's `-Wformat-security` warning for the C `printf`.)

nilsvu commented:
Addressed your comments @wthrowe. @nilsdeppe or @hrueter could you review?

wthrowe commented on src/NumericalAlgorithms/LinearSolver/Convergence.hpp:
```diff
@@ -0,0 +1,188 @@
+ * relative to the start of the first linear solver iteration.
+ * - MaxIterations: Matches if the number of iterations exceeds this limit.
+ *
+ * \note that the smallest possible residual magnitude the linear solver can
```
Start a new sentence after each `\note`.  The rendered documentation does not format in a way that "Note" can be read as part of the sentence.  (It is completely omitted in some cases, even.)

wthrowe reviewed (changes requested):
The new documentation is much clearer.  Aside from the minor formatting comment, I am happy with this.

nilsvu commented:
Thanks for reviewing @wthrowe, squashed and rebased.

nilsvu commented:
ping @nilsdeppe (or forward please)

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Convergence.hpp:
```diff
@@ -0,0 +1,188 @@
+ * artifacts. Requiring an absolute or relative residual below this limit will
+ * likely lead to termination by `MaxIterations`.
+ *
+ * \note Remembere that when the linear operator \f$A\f$ corresponds to a PDE
```
`Remembere` -> `Remember`

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/Convergence.cpp:
```diff
@@ -0,0 +1,149 @@
+      initial_residual_magnitude_(initial_residual_magnitude) {}
+
+ConvergenceReason HasConverged::reason() const noexcept {
+  ASSERT(reason_,
```
I think this should be an `ERROR` because the `*reason_` will throw and I've noticed it's usually quite difficult to figure out what happened when exception is thrown but not caught. I also don't expect that we will pay much in CPU time for changing this to an ERROR and I can see it making somebody's life easier :)

nilsvu replied:
I'm not sure if I understand correctly. The `ASSERT` tests if the `boost::optional<ConvergenceReason> reason_` evaluates to `true`, in which case we are sure that unpacking it with `*` will succeed (and never throw).
I `ASSERT` this because I make clear in the documentation that it is a programming error to call the `reason()` function without first making sure that `HasConverged::bool()` is indeed true.

nilsdeppe replied:
okay

nilsdeppe commented on tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:
```diff
@@ -32,4 +40,38 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearSolver.Tags",
+        "ConvergenceCriteria");
+
+  {
+    INFO("HasConvergedCompute");
```
Can you add a check for `::name()` for this and the compute tag as well please? Thanks!

nilsvu replied:
The `::name()` of the simple tag `HasConverged` is tested above already. The compute tag derives from it, so should I test its `::name()` too?

nilsdeppe replied:
Whatever you prefer

nilsdeppe commented on tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:
```diff
@@ -74,4 +76,13 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearSolver.Tags",
+
+  {
+    INFO("MagnitudeCompute");
+    const auto box =
```
Could you please add a test of `::name()`?

nilsdeppe commented on src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:
```diff
@@ -35,53 +35,65 @@ namespace cg_detail {
+                                     local_residual_square) noexcept {
+          *local_residual_square = residual_square;
         });
+    // Perform a separate `db::mutate` so that we can retrieve the
```
:+1: on this comment :)

nilsdeppe reviewed (changes requested):
A few minor things, I'm happy for you to squash immediately :)

nilsvu commented:
should be done @nilsdeppe @wthrowe

nilsdeppe commented:
Looks like you might have a fun rebase :)

nilsvu committed:
Also observe initial linear solver residuals

Moves the linear solver observe function into a separate file
to make it reusable

nilsvu commented:
It was indeed fun ;)

While rebasing on #1269 I added another commit that fixes some observing (last commit). Could you have a look at that one @nilsdeppe and @wthrowe?

nilsdeppe commented:
@nilsleiffischer IWYU is unhappy

nilsvu commented:
@nilsdeppe @wthrowe fixed iwyu. should be done.

nilsvu commented:
@sxs-collaboration/spectre-core-devs this is ready

nilsdeppe commented:
Travis failures are the usual GCC debug segfaults.
